### DISCUSSION TITLE
Remote cache timeout

### DISCUSSION THREAD

**Discussion by stephenjwatkins** [vercel/turborepo#1993] [Help] [answered]
I just want to make sure I'm understanding current constraints correctly on the remote caching functionality. I have an artifact that's roughly 60mb and I'm attempting to cache it on Vercel from my local machine. It never gets persisted due to hitting a 20s timeout via the Turborepo client. I'm just wanting to make sure that this is expected and that there isn't a way to change the timeout currently?

```
⠴ ...writing to cache...  [0s] 2022-09-17T12:05:28.233-0400 [DEBUG] run: performing request: method=PUT url=https://vercel.com/api/v8/artifacts/238813d6c76dee6d
⠹ ...writing to cache...  [20s] 2022-09-17T12:05:48.234-0400 [ERROR] run: request failed: error="Put \"https://vercel.com/api/v8/artifacts/238813d6c76dee6d\": context deadline exceeded (Client.Timeout exceeded while awaiting headers)" method=PUT url=https://vercel.com/api/v8/artifacts/238813d6c76dee6d
2022-09-17T12:05:48.234-0400 [DEBUG] run: retrying request: request="PUT https://vercel.com/api/v8/artifacts/238813d6c76dee6d" timeout=2s remaining=2
⠸ ...writing to cache...  [42s] 2022-09-17T12:06:10.236-0400 [ERROR] run: request failed: error="Put \"https://vercel.com/api/v8/artifacts/238813d6c76dee6d\": context deadline exceeded (Client.Timeout exceeded while awaiting headers)" method=PUT url=https://vercel.com/api/v8/artifacts/238813d6c76dee6d
2022-09-17T12:06:10.236-0400 [DEBUG] run: retrying request: request="PUT https://vercel.com/api/v8/artifacts/238813d6c76dee6d" timeout=4s remaining=1
⠼ ...writing to cache...  [1m6s] 2022-09-17T12:06:34.238-0400 [ERROR] run: request failed: error="Put \"https://vercel.com/api/v8/artifacts/238813d6c76dee6d\": context deadline exceeded (Client.Timeout exceeded while awaiting headers)" method=PUT url=https://vercel.com/api/v8/artifacts/238813d6c76dee6d
⠴ ...writing to cache...  [1m6s]
```

**Comment by nathanhammond** [comment, marked as answer]
This is possibly a DNS lookup failure? We do not set a custom timeout for asset upload; this has not yet received a response from the server.

Here is the source of the error:
https://github.com/golang/go/blob/f3c39a83a3076eb560c7f687cbb35eef9b506e7d/src/net/http/client.go#L721

All issues I see on the Internet reference a DNS configuration error.

**Reply by stephenjwatkins** [reply]
Interesting. I can poke around with DNS settings to see if there's a difference. I noticed that the 20s timeout was lining up with what appeared to be the same timeout setting in the HTTPClient configuration:

https://github.com/vercel/turborepo/blob/main/cli/internal/client/client.go#L77

**Reply by tm1000** [reply]
Same issue here and I noticed the same 20 second timeout in the code that @stephenjwatkins noticed. Uploading smaller assets work fine (like 1mb) but big assets go past the 20 second timeout

> this has not yet received a response from the server.

Yes. Thats because the asset is being uploaded so there is no response from the server.

The 20 second timeout needs to be configurable

**Reply by tm1000** [reply]
Additionally the timeout was reduced from 60 to 20 here: https://github.com/vercel/turborepo/pull/545
